# Notebook 03 - Price Elasticity

## Summary

In this notebook, we find out price elasticities (PEs) of models against the Flat Price model, and PEs among subsets of premium in a model.

Main findings:
PE of Models against Flat Price model
- Price Elasticity of Champion Model against Flat Price Model: 37.14
- Price Elasticity of Price Test Model against Flat Price Model: 26.80

PE within a model
- Price Test model: highest PE of 1.58 (in magitude) from Q3 (77.52 - 95.76) to Q4 (95.76 - 119.28)
- Champion Model: highest PE of 3.38 (in magitude) from Q3 (83.04 - 90.48) to Q4 (90.48 - 101.88)
- Flat Price model: highest PE of 1.49 (in magitude) from Q2 (58.8 - 81.96) to Q3 (81.96 - 97.2)

---

# Imports

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [2]:
file_path = '/Users/jonwfyeung/Downloads/domestic-and-general'

In [3]:
df_asis = pd.read_csv(file_path + '/data/df_asis.csv')
df_champ = pd.read_csv(file_path + '/data/df_champ.csv')
df_test_cleaned = pd.read_csv(file_path + '/data/df_test_cleaned.csv')

---

# Price Elasticity of Models Against Flat Price Model

Here we find out:
- Price Elasticity of Champion Model against Flat Price Model
- Price Elasticity of Price Test Model against Flat Price Model

using the given formula

$$\frac{\Delta conversion}{\Delta price}=\frac{\frac{(%conversion_{x}-%conversion_{asis})}{%conversion_{asis}}}{\frac{(avg.premium_{x}-avg.premium_{asis})}{avg.premium_{asis}}}$$

In [4]:
# acquiring the corresponding conversion rates and average premium for every model first
df_pe = pd.DataFrame(index=['conversion', 'avg_premium'], columns=['asis', 'champ', 'test'])

df_list = [df_asis, df_champ, df_test_cleaned]

for i, df in enumerate(df_list):
    df_pe.loc['conversion', df_pe.columns[i]] = df['Acc'].mean()
    df_pe.loc['avg_premium', df_pe.columns[i]] = df['Premium'].mean()
    
df_pe

asis      champ       test
conversion    0.291909   0.339676   0.340859
avg_premium  88.138874  88.527168  88.690263

Then, using the conversion rate and the average premium, we can obtain the price elasticities

In [5]:
champ_price_elasticity = ( df_champ['Acc'].mean() / df_asis['Acc'].mean() - 1 )\
                            / ( df_champ['Premium'].mean() / df_asis['Premium'].mean() - 1 )

print(f"Price Elasticity of Champion Model against Flat Price Model:\n{champ_price_elasticity:.2f}")

Price Elasticity of Champion Model against Flat Price Model:
37.14


In [6]:
test_price_elasticity = ( df_test_cleaned['Acc'].mean() / df_asis['Acc'].mean() - 1 )\
                            / ( df_test_cleaned['Premium'].mean() / df_asis['Premium'].mean() - 1 )

print(f"Price Elasticity of Price Test Model against Flat Price Model:\n{test_price_elasticity:.2f}")

Price Elasticity of Price Test Model against Flat Price Model:
26.80


# Price Elasticity Within A Model

I adapt the above formula to obtain the price elasticity(PE) from one subset of premium to another subset of premium within a model, as such:

$$\frac{\Delta conversion}{\Delta price}=\frac{\frac{(%conversion_{i+1}-%conversion_{i})}{%conversion_{i}}}{\frac{(avg.premium_{i+1}-avg.premium_{i})}{avg.premium_{i}}}$$

I first divide the premium into 5 subsets using quantile-cut (5 subsets with roughly equal amounts of data). The 5 subsets are named "Q1", "Q2", "Q3", "Q4" and "Q5" from the lowest range to the highest range.

Then I measure the PE from Q1 to Q2, the PE from Q2 to Q3, the PE from Q3 to Q4, and the PE from Q4 to Q5.

In this way, there is no PE calculated for Q1.

## Price Elasticity in Price Test Model

In [7]:
cut_series, bins_arr = pd.qcut(df_test_cleaned['Premium'], 5, labels=["Q1", "Q2", "Q3", "Q4", "Q5"], retbins=True)

In [8]:
bins_arr

array([ 16.32,  57.6 ,  77.52,  95.76, 119.28, 150.  ])

In [9]:
df_test_cleaned['QCut'] = cut_series
df_test_cleaned.head().T

0                 1  \
Acc                                              0                 1   
ApplianceCode                                   38                67   
ClaimAmount5YearTotal                          0.0            105.48   
ClaimPast5YearCount                              0                 2   
ClaimsToPlansRatio                             0.0              0.25   
ClientAccountDesc                                4                 4   
CurrentDate                             2024-02-05        2023-09-20   
DNISCode                                     CQ9R1             RR3A1   
Fee                                          114.0             150.0   
ManufacturerBrandCode                          233               135   
ManufacturerGuaranteePartsMonths                12               120   
OfferPostcodeSector                          BD229              RM26   
PlanLiveCount                                    2                 3   
PlansAcceptedPast5YearCount                      2                 8   
PlansCancelledPast5YearCount                     0                 7   
Premium                                      94.44            136.32   
PurchasePrice                                379.0             550.0   
SchemeTypeName                                   0                 0   
AgeApplianceDays                               887              1826   
OfferRefNum                        351-042-506-307   118-000-079-955   
PriceDiff                                -0.171579           -0.0912   
PricingVerDesc                    QUO - PRICE TEST  QUO - PRICE TEST   
RelFee                                    0.300792          0.294545   
QCut                                            Q3                Q5   

                                                 2                 3  \
Acc                                              0                 1   
ApplianceCode                                   38                38   
ClaimAmount5YearTotal                          0.0               0.0   
ClaimPast5YearCount                              0                 0   
ClaimsToPlansRatio                             0.0               0.0   
ClientAccountDesc                                4                 4   
CurrentDate                             2023-07-29        2023-10-27   
DNISCode                                     SQ9R1             CQ9R1   
Fee                                          114.0             86.28   
ManufacturerBrandCode                          128               261   
ManufacturerGuaranteePartsMonths                12                12   
OfferPostcodeSector                          WV147              CR00   
PlanLiveCount                                    1                 0   
PlansAcceptedPast5YearCount                      6                 0   
PlansCancelledPast5YearCount                     5                 0   
Premium                                     117.72             66.36   
PurchasePrice                                500.0             529.0   
SchemeTypeName                                   0                 0   
AgeApplianceDays                              1399               122   
OfferRefNum                        351-033-085-284   351-037-427-051   
PriceDiff                                 0.032632         -0.230876   
PricingVerDesc                    QUO - PRICE TEST  QUO - PRICE TEST   
RelFee                                       0.228            0.1631   
QCut                                            Q4                Q2   

                                                 4  
Acc                                              1  
ApplianceCode                                    3  
ClaimAmount5YearTotal                         9.28  
ClaimPast5YearCount                              1  
ClaimsToPlansRatio                             0.0  
ClientAccountDesc                                4  
CurrentDate                             2023-02-06  
DNISCode               

In [10]:
df_test_bins = df_test_cleaned.groupby('QCut').agg({'Acc': 'mean', 'Premium': 'mean'}).T
df_test_bins.rename(index={'Acc': 'conversion', 'Premium': 'avg_premium'}, inplace=True)
df_test_bins

QCut                Q1         Q2         Q3          Q4          Q5
conversion    0.566972   0.406634   0.321056    0.203058    0.205792
avg_premium  43.242862  68.371548  86.570902  106.720807  138.818336

In [11]:
df_test_bins.loc['price_elasticity', 'Q1'] = np.nan

for i in range(4):
    s = df_test_bins.iloc[:,i+1] / df_test_bins.iloc[:,i] - 1
    pe = s['conversion']/s['avg_premium']
    df_test_bins.iloc[2,i+1] = pe

df_test_bins

QCut                     Q1         Q2         Q3          Q4          Q5
conversion         0.566972   0.406634   0.321056    0.203058    0.205792
avg_premium       43.242862  68.371548  86.570902  106.720807  138.818336
price_elasticity        NaN  -0.486654  -0.790639   -1.579036    0.044761

The subsets of premium are:

- Q1: 16.32 - 57.6
- Q2: 57.6 - 77.52
- Q3: 77.52 - 95.76
- Q4: 95.76 - 119.28
- Q5: 119.28 - 150

The PE becomes the highest in magitude (1.58) when the premium range moves from Q3 (77.52 - 95.76) to Q4 (95.76 - 119.28)

## Price Elasticity in Champion Model

In [12]:
cut_series, bins_arr = pd.qcut(df_champ['Premium'], 5, labels=["Q1", "Q2", "Q3", "Q4", "Q5"], retbins=True)

In [13]:
bins_arr

array([ 20.4  ,  74.952,  83.04 ,  90.48 , 101.88 , 150.   ])

In [14]:
df_champ['QCut'] = cut_series
df_champ.head().T

0                1  \
Acc                                             0                0   
ApplianceCode                                   7                3   
ClaimAmount5YearTotal                       84.65              0.0   
ClaimPast5YearCount                             2                0   
ClaimsToPlansRatio                            0.5              0.0   
ClientAccountDesc                               4                4   
CurrentDate                            2024-01-15       2023-09-14   
DNISCode                                    RR4Q1            QQ9R1   
Fee                                          70.8            103.2   
ManufacturerBrandCode                         118              148   
ManufacturerGuaranteePartsMonths               12               12   
OfferPostcodeSector                         RH117            OX495   
PlanLiveCount                                   4                0   
PlansAcceptedPast5YearCount                     4                0   
PlansCancelledPast5YearCount                    7                0   
Premium                                     85.92            88.44   
PrevRTPPriceType                              NaN              NaN   
PurchasePrice                               200.0            300.0   
RTPPriceType                         NEW_BUSINESS     NEW_BUSINESS   
SchemeTypeName                                  6                6   
AgeApplianceDays                             1461             1079   
OfferRefNum                       170-028-720-932  351-035-125-239   
PredConvRate                                 0.28              0.2   
PriceDiff                                0.213559        -0.143023   
PricingVerDesc                      QUO - MODEL Q    QUO - MODEL Q   
RelFee                                      0.354            0.344   
QCut                                           Q3               Q3   

                                                2                3  \
Acc                                             0                0   
ApplianceCode                                  22                6   
ClaimAmount5YearTotal                     1242.91              0.0   
ClaimPast5YearCount                             5                0   
ClaimsToPlansRatio                       0.555556              0.0   
ClientAccountDesc                               4               12   
CurrentDate                            2023-07-28       2024-02-05   
DNISCode                                    QQ9R1            RR4Q1   
Fee                                         110.4            86.52   
ManufacturerBrandCode                         261               30   
ManufacturerGuaranteePartsMonths               12               12   
OfferPostcodeSector                         SA149            RG128   
PlanLiveCount                                   3                0   
PlansAcceptedPast5YearCount                     9                0   
PlansCancelledPast5YearCount                    8                1   
Premium                                    104.28            81.12   
PrevRTPPriceType                              NaN              NaN   
PurchasePrice                               700.0            249.0   
RTPPriceType                         NEW_BUSINESS     NEW_BUSINESS   
SchemeTypeName                                  6                0   
AgeApplianceDays                             1095             2541   
OfferRefNum                       351-033-066-734  170-030-000-275   
PredConvRate                                 0.44             0.02   
PriceDiff                               -0.055435        -0.062413   
PricingVerDesc                      QUO - MODEL Q    QUO - MODEL Q   
RelFee                                   0.157714          0.34747   
QCut                                           Q5               Q2   

                                                4  
Acc                                             0  
ApplianceCode         

In [15]:
df_champ_bins = df_champ.groupby('QCut').agg({'Acc': 'mean', 'Premium': 'mean'}).T
df_champ_bins.rename(index={'Acc': 'conversion', 'Premium': 'avg_premium'}, inplace=True)
df_champ_bins

QCut                Q1         Q2         Q3         Q4          Q5
conversion    0.437608   0.227708   0.262936   0.361077    0.409881
avg_premium  57.426153  80.407251  86.336568  95.860914  122.806087

In [16]:
df_champ_bins.loc['price_elasticity', 'Q1'] = np.nan

for i in range(4):
    s = df_champ_bins.iloc[:,i+1] / df_champ_bins.iloc[:,i] - 1
    pe = s['conversion']/s['avg_premium']
    df_champ_bins.iloc[2,i+1] = pe

df_champ_bins

QCut                     Q1         Q2         Q3         Q4          Q5
conversion         0.437608   0.227708   0.262936   0.361077    0.409881
avg_premium       57.426153  80.407251  86.336568  95.860914  122.806087
price_elasticity        NaN  -1.198577   2.097984   3.383428    0.480858

The subsets of premium are:

- Q1: 20.4 - 74.95
- Q2: 74.95 - 83.04
- Q3: 83.04 - 90.48
- Q4: 90.48 - 101.88
- Q5: 101.88 - 150

The PE becomes the highest in magitude (3.38) when the premium range moves from Q3 (83.04 - 90.48) to Q4 (90.48 - 101.88)

## Price Elasticity in Flat Price Model

In [17]:
cut_series, bins_arr = pd.qcut(df_asis['Premium'], 5, labels=["Q1", "Q2", "Q3", "Q4", "Q5"], retbins=True)

In [18]:
bins_arr

array([ 10.68,  58.8 ,  81.96,  97.2 , 114.  , 150.  ])

In [19]:
df_asis['QCut'] = cut_series
df_asis.head().T

0                1  \
Acc                                             0                0   
ApplianceCode                                   3               38   
ClaimAmount5YearTotal                         0.0          1659.43   
ClaimPast5YearCount                             0               11   
ClaimsToPlansRatio                            0.0         0.647059   
ClientAccountDesc                              13                4   
CurrentDate                            2023-09-04       2023-10-09   
DNISCode                                    SQ301            CQ9R1   
Fee                                         44.28            114.0   
ManufacturerBrandCode                           4              175   
ManufacturerGuaranteePartsMonths               24               12   
OfferPostcodeSector                          BR31             CM96   
PlanLiveCount                                   0               13   
PlansAcceptedPast5YearCount                     0               17   
PlansCancelledPast5YearCount                    0                8   
Premium                                     44.28           119.76   
PrevRTPPriceType                              NaN              NaN   
PurchasePrice                               799.0            750.0   
RTPPriceType                         NEW_BUSINESS     NEW_BUSINESS   
SchemeTypeName                                  2                0   
AgeApplianceDays                               11              707   
OfferRefNum                       108-000-518-493  351-036-478-118   
PredConvRate                                  NaN              NaN   
PriceDiff                                     0.0         0.050526   
PricingVerDesc                     QUO - ASIS FEE   QUO - ASIS FEE   
RelFee                                   0.055419            0.152   
QCut                                           Q1               Q5   

                                                2                3  \
Acc                                             1                0   
ApplianceCode                                  61                7   
ClaimAmount5YearTotal                         0.0           421.66   
ClaimPast5YearCount                             0                4   
ClaimsToPlansRatio                            0.0              1.0   
ClientAccountDesc                               4                4   
CurrentDate                            2023-02-14       2023-06-14   
DNISCode                                    SQ9R1              NaN   
Fee                                          90.0             70.8   
ManufacturerBrandCode                          77               80   
ManufacturerGuaranteePartsMonths               12               12   
OfferPostcodeSector                          CR00             BB50   
PlanLiveCount                                   4                1   
PlansAcceptedPast5YearCount                     4                4   
PlansCancelledPast5YearCount                    0                4   
Premium                                      90.0             70.8   
PrevRTPPriceType                              NaN              NaN   
PurchasePrice                               139.0            250.0   
RTPPriceType                         NEW_BUSINESS     NEW_BUSINESS   
SchemeTypeName                                  0                6   
AgeApplianceDays                              783             1873   
OfferRefNum                       351-025-421-456  034-020-305-517   
PredConvRate                                  NaN              NaN   
PriceDiff                                     0.0              0.0   
PricingVerDesc                     QUO - ASIS FEE   QUO - ASIS FEE   
RelFee                                   0.647482           0.2832   
QCut                                           Q3               Q2   

                                                4  
Acc                                             0  
ApplianceCode         

In [20]:
df_asis_bins = df_asis.groupby('QCut').agg({'Acc': 'mean', 'Premium': 'mean'}).T
df_asis_bins.rename(index={'Acc': 'conversion', 'Premium': 'avg_premium'}, inplace=True)
df_asis_bins

QCut                Q1         Q2         Q3          Q4          Q5
conversion    0.514965   0.340264   0.200341    0.197314    0.200257
avg_premium  44.894859  70.259149  89.596261  105.671590  135.370118

In [21]:
df_asis_bins.loc['price_elasticity', 'Q1'] = np.nan

for i in range(4):
    s = df_asis_bins.iloc[:,i+1] / df_asis_bins.iloc[:,i] - 1
    pe = s['conversion']/s['avg_premium']
    df_asis_bins.iloc[2,i+1] = pe

df_asis_bins

QCut                     Q1         Q2         Q3          Q4          Q5
conversion         0.514965   0.340264   0.200341    0.197314    0.200257
avg_premium       44.894859  70.259149  89.596261  105.671590  135.370118
price_elasticity        NaN  -0.600469  -1.494121   -0.084208    0.053076

The subsets of premium are:

- Q1: 10.68 - 58.8
- Q2: 58.8 - 81.96
- Q3: 81.96 - 97.2
- Q4: 97.2 - 114
- Q5: 114 - 150

The PE becomes the highest in magitude (1.49) when the premium range moves from Q2 (58.8 - 81.96) to Q3 (81.96 - 97.2)